File: frontend/src/api/preferences.ts
import { apiClient } from './client'
import type { OnboardingFormValues } from '../schemas/onboarding'

export type RiskLevel = 'low' | 'medium' | 'high'

export interface OnboardingNotificationPreferences {
  crypto_assets: string[]
  content_types: string[]
}

export interface PreferencesPayload {
  trading_strategy: string
  risk_level?: RiskLevel
  notification_preferences?: OnboardingNotificationPreferences
}

export interface PreferencesResponse extends PreferencesPayload {
  id: number
  user_id: number
  created_at: string
  updated_at: string
}

const INVESTOR_TYPE_TO_RISK_LEVEL: Record<string, RiskLevel> = {
  hodler: 'low',
  day_trader: 'high',
  nft_collector: 'medium',
}

const INVESTOR_TYPE_TO_TRADING_STRATEGY: Record<string, string> = {
  hodler: 'long_term_hold',
  day_trader: 'active_trading',
  nft_collector: 'nft_collecting',
}

export function toPreferencesPayload(values: OnboardingFormValues): PreferencesPayload {
  return {
    trading_strategy: INVESTOR_TYPE_TO_TRADING_STRATEGY[values.investorType] ?? 'balanced_growth',
    risk_level: INVESTOR_TYPE_TO_RISK_LEVEL[values.investorType],
    notification_preferences: {
      crypto_assets: values.cryptoAssets,
      content_types: values.contentTypes,
    },
  }
}

export function submitPreferences(
  payload: PreferencesPayload,
  token: string,
): Promise<PreferencesResponse> {
  return apiClient.post<PreferencesResponse>('/preferences', payload, token)
}

File: backend/schemas/preferences.py
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class PreferencesCreate(BaseModel):
    trading_strategy: str = Field(..., min_length=1, max_length=255)
    risk_level: Optional[RiskLevel] = None
    notification_preferences: Optional[dict] = None


class PreferencesUpdate(BaseModel):
    trading_strategy: Optional[str] = Field(None, min_length=1, max_length=255)
    risk_level: Optional[RiskLevel] = None
    notification_preferences: Optional[dict] = None


class PreferencesResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    trading_strategy: str
    risk_level: Optional[RiskLevel]
    notification_preferences: Optional[dict]
    created_at: datetime
    updated_at: datetime
